RIOT-OS/RIOT
# cpu/esp: improve thread safety in newlib locking functions

### Contribution description

This PR improves the thread safety of the newlib locking functions as following:

- Dynamic allocation and initialization of the mutex used by a newlib locking variable must not be interrupted. Since a thread context switch can occur on exit from an ISR, the allocation and initialization of such a mutex is guarded by disabling interrupts. The same is done for the release of such a mutex.
- Locking functions implicitly allocate and initialize the mutex used by a newlib locking variable if necessary. This implicit allocation and initialization as well as the subsequent acquiring must not be interrupted. Since a thread context switch can occur on exit from an ISR, the implicit allocation and initialization and the subsequent acquiring of such a newlib locking variable must be guarded by disabling interrupts. The same must be done for the release of such a newlib locking variable.
- Module `malloc_thread_safe` is not needed any longer.

This PR fixes issue #18534 and reverts the change made in PR #18535.

### Testing procedure

```
BOARD=esp32-wroom-32 make flash test -C tests/thread_priority_inversion flash test
```
should work with this PR.
```
BOARD=esp32-wroom-32 make-C tests/malloc_thread_safe flash test 
```
should still work.

### Issues/PRs references

Fixes #18534
